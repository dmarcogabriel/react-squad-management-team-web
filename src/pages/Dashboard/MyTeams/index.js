import React, { useState, useEffect } from 'react';
import { MdAdd } from 'react-icons/md';
import { CgArrowsV } from 'react-icons/cg';
import { Link } from 'react-router-dom';
import cn from 'classnames';
import Card from '../../../common/Card';
import SquadItem from './SquadItem';
import { useSquad } from '../../../contexts/Squad';

const MyTeams = ({ className }) => {
  const [squads, setSquads] = useSquad();
  const [myTeams, setMyTeams] = useState([]);

  const handleSelect = (squad) => {
    setMyTeams(
      myTeams.map((team) => ({
        ...team,
        selected: team.id === squad.id,
      }))
    );
  };

  const handleDelete = (squadId) => {
    const newSquads = squads.filter((squad) => squad.id !== squadId);

    setMyTeams(newSquads);
    setSquads(newSquads);
  };

  const handleOrderList = (comparationKey) => {
    const sorted = myTeams.sort((pl1, pl2) =>
      pl1[comparationKey].localeCompare(pl2[comparationKey])
    );

    setMyTeams([...sorted]);
  };

  useEffect(() => {
    if (!myTeams.length) {
      setMyTeams(squads.map((squad) => ({ ...squad, selected: false })));
    }
  }, []);

  return (
    <Card
      dataTestid="container"
      className={className}
      title="My Teams"
      RightComponent={() => (
        <Link
          to="create"
          className={cn(
            'bg-gradient-to-t from-secondary-dark',
            ' to-secondary-light text-white p-1 rounded-lg'
          )}
        >
          <MdAdd size={22} />
        </Link>
      )}
    >
      <table className="w-full">
        <thead>
          <tr className="text-base flex">
            <th>
              <button
                data-testid="order-by-name"
                type="button"
                onClick={() => handleOrderList('name')}
                className={cn(
                  'flex items-center mx-3 py-2',
                  'border-r-2 border-gray-100'
                )}
              >
                <p className="font-normal mr-5">Name</p>

                <CgArrowsV className="mx-2" />
              </button>
            </th>
            <th>
              <button
                data-testid="order-by-description"
                type="button"
                onClick={() => handleOrderList('description')}
                className="flex items-center w-full mx-3 py-2"
              >
                <p className="font-normal mr-5">Description</p>

                <CgArrowsV className="mx-2" />
              </button>
            </th>
          </tr>
        </thead>

        <tbody data-testid="squad-list">
          {myTeams.map((squad, i) => (
            <SquadItem
              dataTestid={`squad-${i}`}
              squadSelectButtonTestid={`squad-select-${i}`}
              squadDeleteButtonTestid={`squad-delete-${i}`}
              key={squad.id}
              squad={squad}
              onClick={handleSelect}
              onDelete={handleDelete}
            />
          ))}
        </tbody>
      </table>
    </Card>
  );
};

export default MyTeams;
