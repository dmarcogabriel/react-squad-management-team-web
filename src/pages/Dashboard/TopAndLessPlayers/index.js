import React, { useState, useEffect } from 'react';
import lodash from 'lodash';
import { usePlayers } from '../../../contexts/Players';
import Content from './Content';

const TopAndLessPlayers = () => {
  const [mostPickedPlayer, setMostPickedPlayer] = useState(null);
  const [lessPickedPlayer, setLessPickedPlayer] = useState(null);

  const [players] = usePlayers();

  useEffect(() => {
    if (players && players.length) {
      const allPlayers = lodash.groupBy(players, 'player_id');

      const a = Object.keys(allPlayers).map((k) => allPlayers[k]);

      const [mostPicked] = a.reduce((b, c) => (b.length > c.length ? b : c));

      const [lessPicked] = a.reduce((b, c) => (b.length > c.length ? c : b));

      setMostPickedPlayer(mostPicked);
      setLessPickedPlayer(lessPicked);
    }
  }, []);

  return (
    <div className="rounded-lg flex flex-col md:flex-row shadow-lg bg-gradient-to-t from-secondary-dark to-secondary-light text-white mb-5">
      <Content
        className="border-b-2 md:border-r-2 md:border-b-0 border-opacity-25 border-white"
        playerName={mostPickedPlayer && mostPickedPlayer.display_name}
        percentage="0%"
        title="Most picked player"
      />

      <Content
        playerName={lessPickedPlayer && lessPickedPlayer.display_name}
        percentage="0%"
        title="Less picked player"
      />
    </div>
  );
};

export default TopAndLessPlayers;
