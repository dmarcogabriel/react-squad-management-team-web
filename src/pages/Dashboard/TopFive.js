import React from 'react';
import Card from '../../common/Card';

const TopFive = () => {
  return (
    <Card title="Top 5">
      <div className="flex flex-col md:flex-row mx-2">
        <div className="w-full mb-5 md:mr-8">
          <h2 className="text-base mb-3">Highest avg age</h2>

          <div className="bg-background rounded-lg p-1">
            {Array.from(Array(5)).map((_, i) => (
              <div
                key={String(i)}
                className="flex items-center justify-between px-4 py-2 my-1 bg-white rounded-lg"
              >
                <p className="text-sm">Arsenal FC</p>
                <p className="font-bold text-sm">31.2</p>
              </div>
            ))}
          </div>
        </div>

        <div className="w-full">
          <h2 className="text-base mb-3">Lowest avg age</h2>

          <div className="bg-background rounded-lg p-1">
            {Array.from(Array(5)).map((_, i) => (
              <div
                key={String(i)}
                className="flex items-center justify-between px-4 py-2 my-1 bg-white rounded-lg"
              >
                <p className="text-sm">Arsenal FC</p>
                <p className="font-bold text-sm">31.2</p>
              </div>
            ))}
          </div>
        </div>
      </div>
    </Card>
  );
};

export default TopFive;
