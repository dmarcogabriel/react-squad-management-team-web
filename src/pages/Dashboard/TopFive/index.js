import React, { useEffect, useState } from 'react';
import Card from '../../../common/Card';
import { useSquad } from '../../../contexts/Squad';

const TopFive = () => {
  const [highestAvgAge, setHighestAvgAge] = useState([]);
  const [lowestAvgAge, setLowestAvgAge] = useState([]);

  const [squads] = useSquad();

  const getPlayersAverage = () => {
    const lowestAvg = squads
      .sort((sq1, sq2) => sq1.ageAvg - sq2.ageAvg)
      .slice(0, 5);

    const highestAvg = squads
      .sort((sq1, sq2) => sq1.ageAvg - sq2.ageAvg)
      .reverse()
      .slice(0, 5);

    setHighestAvgAge(highestAvg);
    setLowestAvgAge(lowestAvg);
  };

  useEffect(() => {
    getPlayersAverage();
  }, []);

  return (
    <Card title="Top 5">
      <div className="flex flex-col md:flex-row mx-2">
        <div className="w-full mb-5 md:mr-8">
          <h2 className="text-base mb-3">Highest avg age</h2>

          <div
            data-testid="highest-avgs"
            className="bg-background rounded-lg p-1"
          >
            {highestAvgAge.map((squad) => (
              <div
                key={squad.id}
                className="flex items-center justify-between px-4 py-2 my-1 bg-white rounded-lg"
              >
                <p data-testid={`highest-${squad.id}-name`} className="text-sm">
                  {squad.name}
                </p>
                <p className="font-bold text-sm">{squad.ageAvg.toFixed(1)}</p>
              </div>
            ))}
          </div>
        </div>

        <div className="w-full">
          <h2 className="text-base mb-3">Lowest avg age</h2>

          <div
            data-testid="lowest-avgs"
            className="bg-background rounded-lg p-1"
          >
            {lowestAvgAge.map((squad) => (
              <div
                key={squad.id}
                className="flex items-center justify-between px-4 py-2 my-1 bg-white rounded-lg"
              >
                <p data-testid={`lowest-${squad.id}-name`} className="text-sm">
                  {squad.name}
                </p>
                <p className="font-bold text-sm">{squad.ageAvg.toFixed(1)}</p>
              </div>
            ))}
          </div>
        </div>
      </div>
    </Card>
  );
};

export default TopFive;
