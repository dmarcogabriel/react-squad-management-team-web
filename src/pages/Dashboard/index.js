import React from 'react';
import Page from '../../common/Page';
import MyTeams from './MyTeams';
import TopAndLessPlayers from './TopAndLessPlayers';
import TopFive from './TopFive';

const Dashboard = () => {
  return (
    <Page>
      <MyTeams className="md:mr-10" />

      <div className="flex flex-col w-full">
        <TopFive />

        <TopAndLessPlayers />
      </div>
    </Page>
  );
};

export default Dashboard;
