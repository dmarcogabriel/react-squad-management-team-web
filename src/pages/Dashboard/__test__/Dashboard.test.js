import { cleanup } from '@testing-library/react';
import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Dashboard from '..';

const Component = () => (
  <Router>
    <Dashboard />
  </Router>
);

describe('<Dashboard />', () => {
  afterEach(cleanup);

  it('renders Card without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Component />, div);
  });

  it('match snapshot', () => {
    expect(renderer.create(<Component />).toJSON()).toMatchSnapshot();
  });
});
