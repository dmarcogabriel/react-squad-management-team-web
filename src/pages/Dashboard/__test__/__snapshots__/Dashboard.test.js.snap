// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> match snapshot 1`] = `
Array [
  <nav
    className="flex px-8 py-2 items-center justify-between bg-gradient-to-r from-primary-dark to-secondary-dark text-white"
    data-testid="header"
  >
    <div>
      <p
        className="text-lg"
      >
        Squad Management Tool
      </p>
    </div>
    <div
      className="flex items-center"
    >
      <p
        className="text-sm mr-2 invisible md:visible"
      >
        John Doe
      </p>
      <p
        className="bg-white py-2 px-3 rounded-full text-black"
      >
        JD
      </p>
    </div>
  </nav>,
  <div
    className="flex flex-col md:flex-row bg-background justify-center w-full px-5 md:px-10"
    data-testid="page"
  >
    <div
      className="bg-white rounded-lg shadow-lg my-10 w-full md:mr-10"
      data-testid="container"
    >
      <div
        className="flex p-8 justify-between items-center border-b-2 border-gray-200"
      >
        <h1
          className="text-secondary-dark font-bold text-md"
          data-testid="card-title"
        >
          My Teams
        </h1>
        <a
          className="bg-gradient-to-t from-secondary-dark to-secondary-light text-white p-1 rounded-lg"
          href="/create"
          onClick={[Function]}
        >
          <svg
            fill="currentColor"
            height={22}
            size={22}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width={22}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </a>
      </div>
      <div
        className="py-5 px-2"
        data-testid="card-children"
      >
        <table
          className="w-full"
        >
          <thead>
            <tr
              className="text-base flex"
            >
              <th>
                <button
                  className="flex items-center mx-3 py-2 border-r-2 border-gray-200"
                  data-testid="order-by-name"
                  onClick={[Function]}
                  type="button"
                >
                  <p
                    className="font-normal mr-5"
                  >
                    Name
                  </p>
                  <svg
                    className="mx-2"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7.75739 5.04077L9.1716 6.45498L11.0001 4.62652V10H13.0001V4.62661L14.8284 6.45498L16.2426 5.04077L12 0.798126L7.75739 5.04077Z"
                      fill="currentColor"
                    />
                    <path
                      d="M16.2426 18.9593L14.8284 17.545L13.0001 19.3734V14H11.0001V19.3735L9.1716 17.545L7.75739 18.9593L12 23.2019L16.2426 18.9593Z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </th>
              <th>
                <button
                  className="flex items-center w-full mx-3 py-2"
                  data-testid="order-by-description"
                  onClick={[Function]}
                  type="button"
                >
                  <p
                    className="font-normal mr-5"
                  >
                    Description
                  </p>
                  <svg
                    className="mx-2"
                    fill="none"
                    height="1em"
                    stroke="currentColor"
                    strokeWidth="0"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7.75739 5.04077L9.1716 6.45498L11.0001 4.62652V10H13.0001V4.62661L14.8284 6.45498L16.2426 5.04077L12 0.798126L7.75739 5.04077Z"
                      fill="currentColor"
                    />
                    <path
                      d="M16.2426 18.9593L14.8284 17.545L13.0001 19.3734V14H11.0001V19.3735L9.1716 17.545L7.75739 18.9593L12 23.2019L16.2426 18.9593Z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </th>
            </tr>
          </thead>
          <tbody
            data-testid="squad-list"
          />
        </table>
      </div>
    </div>
    <div
      className="flex flex-col w-full"
    >
      <div
        className="bg-white rounded-lg shadow-lg my-10 w-full"
        data-testid="card"
      >
        <div
          className="flex p-8 justify-between items-center border-b-2 border-gray-200"
        >
          <h1
            className="text-secondary-dark font-bold text-md"
            data-testid="card-title"
          >
            Top 5
          </h1>
        </div>
        <div
          className="py-5 px-2"
          data-testid="card-children"
        >
          <div
            className="flex flex-col md:flex-row mx-2"
          >
            <div
              className="w-full mb-5 md:mr-8"
            >
              <h2
                className="text-base mb-3"
              >
                Highest avg age
              </h2>
              <div
                className="bg-background rounded-lg p-1"
                data-testid="highest-avgs"
              />
            </div>
            <div
              className="w-full"
            >
              <h2
                className="text-base mb-3"
              >
                Lowest avg age
              </h2>
              <div
                className="bg-background rounded-lg p-1"
                data-testid="lowest-avgs"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="rounded-lg flex flex-col md:flex-row shadow-lg bg-gradient-to-t from-secondary-dark to-secondary-light text-white mb-5"
      >
        <div
          className="w-full p-5 border-b-2 md:border-r-2 md:border-b-0 border-opacity-25 border-white"
          data-testid="most-picked"
        >
          <h2
            className="font-bold text-lg text-center"
            data-testid="title"
          >
            Most picked player
          </h2>
          <div
            className="relative flex justify-center my-3"
          >
            <div
              className="relative p-12 border-dashed border-2 text-primary-dark rounded-full overflow-hidden"
            >
              <img
                alt="Player Avatar"
                className="absolute right-0 top-0 w-full h-full"
                src="playerAvatarPlaceholder.png"
              />
            </div>
            <p
              className="absolute right-0 top-0 text-base border-b-2 border-white pr-5"
              data-testid="percentage"
            >
              0%
            </p>
          </div>
        </div>
        <div
          className="w-full p-5"
          data-testid="less-picked"
        >
          <h2
            className="font-bold text-lg text-center"
            data-testid="title"
          >
            Less picked player
          </h2>
          <div
            className="relative flex justify-center my-3"
          >
            <div
              className="relative p-12 border-dashed border-2 text-primary-dark rounded-full overflow-hidden"
            >
              <img
                alt="Player Avatar"
                className="absolute right-0 top-0 w-full h-full"
                src="playerAvatarPlaceholder.png"
              />
            </div>
            <p
              className="absolute right-0 top-0 text-base border-b-2 border-white pr-5"
              data-testid="percentage"
            >
              0%
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bg-footer py-3 flex justify-center"
    data-testid="footer"
  >
    <p
      className="text-sm text-gray-900"
      data-testid="footer-text"
    >
      2020 - All rights reserved
    </p>
  </div>,
]
`;
